### PR TITLE
Add FilteringSpanExporter with composable SpanFilter and TraceFilter interfaces

### PR DESCRIPTION
**Pull request by udaysagar2177**
## Summary

Adds a trace-level filtering `SpanExporter` to the `processors` module. `FilteringSpanExporter` wraps a delegate exporter and only forwards spans belonging to traces that match at least one configured filter. Filtering is trace-aware: if any filter matches, all spans sharing that trace ID are exported together.

Two filter interfaces enable composable, user-extensible filtering:

- **`SpanFilter`** — per-span evaluation (e.g., error status, slow duration)
- **`TraceFilter`** — batch-level evaluation over all spans sharing a trace ID (e.g., overall trace wall-clock duration)

Built-in implementations:

| Class | Type | Behavior |
|-------|------|----------|
| `ErrorSpanFilter` | `SpanFilter` | Keeps traces containing any span with `StatusCode.ERROR` |
| `DurationSpanFilter` | `SpanFilter` | Keeps traces containing any span exceeding a duration threshold |
| `TraceDurationFilter` | `TraceFilter` | Keeps traces whose wall-clock duration (max end - min start) exceeds a threshold |

An optional `Meter` parameter emits a `filtering.span.exporter.dropped` counter with a `reason` attribute for observability into dropped spans.

### Motivation

Services with heavy sampling often want to prioritize exporting traces that contain errors or exhibit unusual latency, without increasing overall sampling rate. This exporter wrapper provides that capability as a composable, upstream component that any OTel Java user can plug into their SDK configuration.

The existing `InterceptableSpanExporter` in this module operates per-span via the `Interceptor` interface, but has no trace-level grouping. `FilteringSpanExporter` fills this gap by grouping spans by trace ID and making keep/drop decisions at the trace level.

### Usage

```java
SpanExporter filtered = new FilteringSpanExporter(
    delegate,
    Arrays.asList(new ErrorSpanFilter(), new DurationSpanFilter(2000)),
    Collections.singletonList(new TraceDurationFilter(10000)));

// Custom filters via lambda
SpanFilter nameFilter = span -> span.getName().contains("important");

// With metrics
SpanExporter withMetrics = new FilteringSpanExporter(
    delegate, spanFilters, traceFilters, meter);
```

## Changes

| File | Description |
|------|-------------|
| `SpanFilter.java` | Interface for per-span filtering |
| `TraceFilter.java` | Interface for batch-level trace filtering |
| `ErrorSpanFilter.java` | Built-in: keeps traces with error spans |
| `DurationSpanFilter.java` | Built-in: keeps traces with slow spans |
| `TraceDurationFilter.java` | Built-in: keeps traces with long wall-clock duration |
| `FilteringSpanExporter.java` | Core exporter wrapper with single-pass filtering |
| `ErrorSpanFilterTest.java` | 3 tests |
| `DurationSpanFilterTest.java` | 3 tests |
| `TraceDurationFilterTest.java` | 5 tests |
| `FilteringSpanExporterTest.java` | 13 tests (composition, grouping, metrics, delegation) |
| `README.md` | Documentation |

## Test plan

- [x] `ErrorSpanFilterTest` — error, OK, UNSET status codes
- [x] `DurationSpanFilterTest` — over/at/under threshold boundary cases
- [x] `TraceDurationFilterTest` — over/at/under threshold, single-span trace
- [x] `FilteringSpanExporterTest` — trace grouping, mixed traces, span-only/trace-only/custom filter composition, dropped-span metrics, flush/shutdown delegation, empty batch
- [x] Spotless formatting passes
- [x] Compiles with NullAway and Error Prone checks

### CODE REVIEW

File: processors/README.md [hunk 1]
```diff
@@ -32,6 +32,47 @@ logger_provider:
...
+
+- `SpanFilter` - evaluates individual spans (e.g., error status, slow duration)
+- `TraceFilter` - evaluates all spans belonging to a trace within the batch (e.g., overall trace wall-clock duration)
+
+A trace is kept if any `SpanFilter` matches any of its spans, OR any `TraceFilter` matches the trace's span group. All spans sharing a trace ID are exported together.
+
+Built-in filters:
+- `ErrorSpanFilter` - keeps traces containing any span with error status
+- `DurationSpanFilter` - keeps traces containing any span exceeding a duration threshold
+- `TraceDurationFilter` - keeps traces whose wall-clock duration (max end - min start) exceeds a threshold
```
Comment (breedx-splk) [inline processors/README.md]: (this should fix the linter)

```suggestion
Built-in filters:

- `ErrorSpanFilter` - keeps traces containing any span with error status
- `DurationSpanFilter` - keeps traces containing any span exceeding a duration threshold
- `TraceDurationFilter` - keeps traces whose wall-clock duration (max end - min start) exceeds a threshold
```

Comment (udaysagar2177) [reply processors/README.md]: Done!

**Review by LikeTheSalad — APPROVED** [review]
LGTM! Thank you 🙏

File: processors/src/main/java/io/opentelemetry/contrib/filter/TraceDurationFilter.java [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+   *
+   * @param thresholdMs the trace duration threshold in milliseconds; traces with wall-clock
+   *     duration strictly greater than this are considered interesting
+   */
+  public TraceDurationFilter(long thresholdMs) {
+    this.thresholdNanos = TimeUnit.MILLISECONDS.toNanos(thresholdMs);
+  }
+
+  @Override
+  public boolean shouldKeep(String traceId, Collection<SpanData> spans) {
```
Comment (jaydeluca) [inline processors/src/main/java/io/opentelemetry/contrib/filter/TraceDurationFilter.java:34]:
```suggestion
  public boolean shouldKeep(String traceId, Collection<SpanData> spans) {
    if (spans.isEmpty()) {
      return false;
    }
```

we can add an early return for an empty list of spans otherwise you end up with `maxEnd - minStart = Long.MIN_VALUE - Long.MAX_VALUE = 1` down on line 42 and it'll return true.

Related test that currently would fail:

```java
  @Test
  void worksWithEmptyList() {
    TraceDurationFilter filter = new TraceDurationFilter(0);
    assertThat(filter.shouldKeep("trace-1", emptyList())).isFalse();
  }
```

File: processors/src/main/java/io/opentelemetry/contrib/filter/FilteringSpanExporter.java [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+   *     disable metrics
+   */
+  public FilteringSpanExporter(
+      SpanExporter delegate,
+      List<SpanFilter> spanFilters,
+      List<TraceFilter> traceFilters,
+      @Nullable Meter meter) {
+    this.delegate = delegate;
+    this.spanFilters = Collections.unmodifiableList(new ArrayList<>(spanFilters));
+    this.traceFilters = Collections.unmodifiableList(new ArrayList<>(traceFilters));
```
Comment (jaydeluca) [inline processors/src/main/java/io/opentelemetry/contrib/filter/FilteringSpanExporter.java]: should we do any kind of validation here?

```suggestion
    this.delegate = Objects.requireNonNull(delegate, "delegate");
    Objects.requireNonNull(spanFilters, "spanFilters");
    Objects.requireNonNull(traceFilters, "traceFilters");
    this.spanFilters = Collections.unmodifiableList(new ArrayList<>(spanFilters));
    this.traceFilters = Collections.unmodifiableList(new ArrayList<>(traceFilters));
```

File: processors/src/main/java/io/opentelemetry/contrib/filter/TraceDurationFilter.java [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+  private final long thresholdNanos;
+
+  /**
+   * Creates a new {@code TraceDurationFilter}.
+   *
+   * @param thresholdMs the trace duration threshold in milliseconds; traces with wall-clock
+   *     duration strictly greater than this are considered interesting
+   */
+  public TraceDurationFilter(long thresholdMs) {
+    this.thresholdNanos = TimeUnit.MILLISECONDS.toNanos(thresholdMs);
```
Comment (jaydeluca) [inline processors/src/main/java/io/opentelemetry/contrib/filter/TraceDurationFilter.java]:
```suggestion
    if (thresholdMs < 0) {
      throw new IllegalArgumentException("thresholdMs must be non-negative, got: " + thresholdMs);
    }
    this.thresholdNanos = TimeUnit.MILLISECONDS.toNanos(thresholdMs);
```

Also, same comment as `DurationSpanFilter` about using `Duration` instead of long

File: processors/src/main/java/io/opentelemetry/contrib/filter/DurationSpanFilter.java [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+  private final long thresholdNanos;
+
+  /**
+   * Creates a new {@code DurationSpanFilter}.
+   *
+   * @param thresholdMs the duration threshold in milliseconds; spans with duration strictly greater
+   *     than this are considered interesting
+   */
+  public DurationSpanFilter(long thresholdMs) {
+    this.thresholdNanos = TimeUnit.MILLISECONDS.toNanos(thresholdMs);
```
Comment (jaydeluca) [inline processors/src/main/java/io/opentelemetry/contrib/filter/DurationSpanFilter.java]: we should validate the input here

```suggestion
    if (thresholdMs < 0) {
      throw new IllegalArgumentException("thresholdMs must be non-negative, got: " + thresholdMs);
    }
    this.thresholdNanos = TimeUnit.MILLISECONDS.toNanos(thresholdMs);
```

Also, perhaps it would make the API a bit more intuitive if we accepted a Duration instead of a long? `new DurationSpanFilter(Duration.ofSeconds(2))` feels more intuitive than `new DurationSpanFilter(2000)`

File: processors/src/main/java/io/opentelemetry/contrib/filter/FilteringSpanExporter.java [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+    if (meter != null) {
+      this.droppedSpansCounter =
+          meter
+              .counterBuilder("filtering.span.exporter.dropped")
```
Comment (jaydeluca) [inline processors/src/main/java/io/opentelemetry/contrib/filter/FilteringSpanExporter.java]: in the [sdk](https://opentelemetry.io/docs/specs/semconv/otel/sdk-metrics/#metric-otelsdkexporterspanexported) they have metrics that look like `otel.sdk.exporter.span.exported` and `otel.sdk.processor.span.processed`, so perhaps we should use something closer to that pattern like `otel.contrib.exporter.span.filtered` or `otel.contrib.processor.span.dropped`?  Or maybe having `contrib` doesn't make sense in this context

@trask  wdyt?

```suggestion
              .counterBuilder("otel.contrib.processor.span.filtered")
```

File: processors/src/main/java/io/opentelemetry/contrib/filter/FilteringSpanExporter.java [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+              .setDescription("Number of spans dropped by the filtering span exporter")
```
Comment (jaydeluca) [inline processors/src/main/java/io/opentelemetry/contrib/filter/FilteringSpanExporter.java:79]:
```suggestion
              .setDescription("Number of spans dropped by the filtering span exporter")
              .setUnit("{span}")
```

File: processors/src/main/java/io/opentelemetry/contrib/filter/TraceDurationFilter.java (hunk 2 above)
Comment (udaysagar2177) [reply processors/src/main/java/io/opentelemetry/contrib/filter/TraceDurationFilter.java:34]: Good catch on the overflow — fixed with an early return. Added a test for this case too.

File: processors/src/main/java/io/opentelemetry/contrib/filter/FilteringSpanExporter.java (hunk 3 above)
Comment (udaysagar2177) [reply processors/src/main/java/io/opentelemetry/contrib/filter/FilteringSpanExporter.java]: Done, added `Objects.requireNonNull` for all three parameters.

File: processors/src/main/java/io/opentelemetry/contrib/filter/TraceDurationFilter.java (hunk 4 above)
Comment (udaysagar2177) [reply processors/src/main/java/io/opentelemetry/contrib/filter/TraceDurationFilter.java]: Added both — validation and switched to `Duration`. Agreed it reads much better: `new TraceDurationFilter(Duration.ofSeconds(10))`.

File: processors/src/main/java/io/opentelemetry/contrib/filter/DurationSpanFilter.java (hunk 5 above)
Comment (udaysagar2177) [reply processors/src/main/java/io/opentelemetry/contrib/filter/DurationSpanFilter.java]: Done on both — validation added and switched to `Duration`. `new DurationSpanFilter(Duration.ofSeconds(2))` is much more intuitive.

File: processors/src/main/java/io/opentelemetry/contrib/filter/FilteringSpanExporter.java (hunk 6 above)
Comment (udaysagar2177) [reply processors/src/main/java/io/opentelemetry/contrib/filter/FilteringSpanExporter.java]: Used `otel.contrib.processor.span.filtered` — aligns well with the SDK naming conventions. Happy to adjust if @trask has a different preference.

File: processors/src/main/java/io/opentelemetry/contrib/filter/FilteringSpanExporter.java (hunk 7 above)
Comment (udaysagar2177) [reply processors/src/main/java/io/opentelemetry/contrib/filter/FilteringSpanExporter.java:79]: Done, added `.setUnit("{span}")`.

**Comment by jaydeluca** [comment]
@udaysagar2177 looks like there's some formatting issues, could you run spotless to fix that up?

**Review by breedx-splk — APPROVED** [review]
